bpo-35934: Add socket.create_server() utility function

giampaolo opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-35934](https://bugs.python.org/issue35934) -->
https://bugs.python.org/issue35934
<!-- /issue-number -->

giampaolo committed:
add socket.bind_socket() function + tests

giampaolo committed:
make ftplib use bind_socket() (reuse code)

giampaolo committed:
use bind_socket() in tests (reuse code)

giampaolo committed:
set default backlog to 128; provide a more informative message if SO_REUSEPORT is not supported

giampaolo committed:
add 'flags' parameter

giampaolo committed:
use bind_socket() in more unit-tests

giampaolo committed:
add comment

giampaolo committed:
add NEWS entry

asvetlov commented on Doc/library/socket.rst:
```diff
@@ -595,6 +595,21 @@ The following functions all create :ref:`socket objects <socket-objects>`.
    .. versionchanged:: 3.2
       *source_address* was added.
 
+.. function:: bind_socket(address, family=AF_UNSPEC, type=SOCK_STREAM, *,
```
Would you consider making all arguments except `address` keyword-only?

giampaolo replied:
Mmm I have mixed feelings about this. Basically I did it like this in order to stay consistent with the signature of `socket.socket()`, `socket.socketpair()` and more marginally `socket.getaddrinfo()`.

giampaolo replied:
I changed my mind about this (and updated code).

asvetlov commented on Lib/socket.py:
```diff
@@ -728,6 +728,78 @@ def create_connection(address, timeout=_GLOBAL_DEFAULT_TIMEOUT,
+
+    If *host* is an empty string or None all network interfaces are assumed.
+
+    If *family* is AF_UNSPEC or None the address family will be
```
I think better to don't support `None` for `family`, `AF_UNSPEC` should be enough.
In most cases user will not pass the argument for default behavior.
You don't support `None` for `type` for example.

asvetlov replied:
BTW, `AF_UNSPEC` == `0`, pretty reasonable default value.

giampaolo replied:
Agreed (not supporting None).

asvetlov commented on Lib/socket.py:
```diff
@@ -728,6 +728,78 @@ def create_connection(address, timeout=_GLOBAL_DEFAULT_TIMEOUT,
+    if host == "":
+        # https://mail.python.org/pipermail/python-ideas/2013-March/019937.html
+        host = None  # all interfaces
+    if not family:
```
Technically `[]` and `0.0` are acceptable.
I don't insist on argument type check in runtime here, avoiding `None` would be enough.

giampaolo replied:
+1

asvetlov commented on Lib/socket.py:
```diff
@@ -728,6 +728,78 @@ def create_connection(address, timeout=_GLOBAL_DEFAULT_TIMEOUT,
+        host = None  # all interfaces
+    if not family:
+        family = AF_UNSPEC
+    reuse_addr = hasattr(_socket, 'SO_REUSEADDR') and \
```
`asyncio` has this logic only if `reuse_addr` parameter is `None` (default).
It allows to explicitly enforce and *disable* reuse addr feature.
I've found the logic useful for some cases.

giampaolo replied:
Agreed. I added a `reuse_addr` parameter.

njsmith replied:
@asvetlov I'm *extremely* curious what situation you've found where you don't want `SO_REUSEADDR` :-).

As far as I understand it, `SO_REUSEADDR` is simply always wanted on POSIX; it's a classic "unbreak-me" flag. I don't think there's any harm in allowing people to disable it, I'm just really curious to hear why anyone would want to do this :-).

On Windows, I have a stronger opinion: we should not provide any way to control `SO_REUSEADDR` on Windows. On Windows, the behavior that most people expect `SO_REUSEADDR` to control (i.e., binding to ports that are in the TIME_WAIT state) is always enabled, and cannot be configured. But if you use `SO_REUSEADDR` anyway, then this creates a security bug (and also a regular bug). It would be irresponsible to provide a flag where `False` does nothing and `True` breaks security. Simply making the default `False` on Windows is not enough; using `SO_REUSEADDR` should be impossible.

On Windows you should also enable `SO_EXCLUSIVEADDRUSE`, which is required to fully fix the security bug I mentioned. It has no other effect except to fix this bug.

giampaolo replied:
As for `SO_REUSEADDR` this SO reply explains why disabling `SO_REUSEADDR` can be desirable (although not frequently): https://stackoverflow.com/a/14388707. In short, when not set it results in `EADDRINUSE` error if:
1) a wildcard address ("all" interfaces) is specified as "host" and some address/port is already bound to another socket
2) there is some pending data to send

As such it appears the argument/toggle is desirable.

As for `SO_REUSEADDR` it seems the semantic is different on Windows. When set to `1` it forces the socket to forcibly bind to a port in use by another socket, whereas on UNIX this results in `EADDRINUSE` error. It also sets the process in a state where it'll no longer respond to any signals or graceful kills, and must be forcibly killed (see [test.support](https://github.com/python/cpython/blob/7ab3d1573c123fdd582a239d01f8475651df38f2/Lib/test/support/__init__.py#L652) and [BPO-2550](https://bugs.python.org/issue2550)). So I agree `reuse_addr=True` should be ignored on Windows (and documented).

As for `SO_EXCLUSIVEADDRUSE` (Windows), I'm dubious. [MS doc](http://msdn2.microsoft.com/en-us/library/ms740621(VS.85).aspx) states:

>  a socket with the SO_EXCLUSIVEADDRUSE set cannot necessarily be reused immediately after socket closure. For example, if a listening socket with SO_EXCLUSIVEADDRUSE set accepts a connection and is then subsequently closed, another socket (also with SO_EXCLUSIVEADDRUSE) cannot bind to the same port as the first socket until the original connection becomes inactive.

...so it seems this is not desirable to set by default or change the meaning of `reuse_addr` opt on Windows to mean `SO_EXCLUSIVEADDRUSE` instead of `SO_REUSEADDR`. Also, I took a peek at Tornado, Twisted, asyncio, curio and none of them use `SO_EXCLUSIVEADDRUSE`.

njsmith replied:
Yes, I'm pretty sure those MS docs are why none of the other libraries use it. However, the docs are wrong :-). [Try it](https://github.com/python-trio/trio/blob/master/notes-to-self/time-wait-windows-exclusiveaddruse.py) – Windows is totally happy to let you `bind` a new listening socket to a port that still has live connections from a previous listening socket.

giampaolo replied:
I wouldn't be so sure about [this script](https://github.com/python-trio/trio/blob/master/notes-to-self/time-wait-windows-exclusiveaddruse.py). It appears the "rebind with TIME_WAIT socket" test succeeds regardless of `SO_EXCLUSIVEADDRUSE` (I tried).

The whole point of `reuse_addr=True` on UNIX is to avoid getting `EADDRINUSE` when the previous socket is still in `TIME_WAIT` state. AFAIU WinSock behaves like that already without setting any option. For this reason I think it's correct to not set `SO_REUSEADDR` by default on Windows, nor to map `reuse_addr=True` to `SO_EXCLUSIVEADDRUSE` because fundamentally `SO_REUSEADDR` and `SO_EXCLUSIVEADDRUSE` have different meanings.

Also I think I changed my mind about disallowing to set `SO_REUSEADDR` on Windows because despite the security implications it appears there is a use-case for it. From MS doc:

> (about SO_REUSERADDR) The exception to this non-deterministic behavior is multicast sockets. If two sockets are bound to the same interface and port and are members of the same multicast group, data will be delivered to both sockets, rather than an arbitrarily chosen one.
> [...]
> The SO_REUSEADDR option has very few uses in normal applications aside from multicast sockets where data is delivered to all of the sockets bound on the same port.

As such I prefer the user to be able to set/unset it if he/she wants to.

njsmith replied:
> I wouldn't be so sure about this script. It appears the "rebind with TIME_WAIT socket" test succeeds regardless of SO_EXCLUSIVEADDRUSE (I tried).

Dunno why you're "not so sure" about the script. It appears to have done exactly what I claimed.

> For this reason I think it's correct to not set SO_REUSEADDR by default on Windows, nor to map reuse_addr=True to SO_EXCLUSIVEADDRUSE because fundamentally SO_REUSEADDR and SO_EXCLUSIVEADDRUSE have different meanings.

I agree. That's why I never suggested doing this. This is a very odd conversation.

Again, my suggestion is: on Windows, use SO_EXCLUSIVEADDRUSE and don't use SO_REUSEADDR. That's all.

> Also I think I changed my mind about disallowing to set SO_REUSEADDR on Windows because despite the security implications it appears there is a use-case for it.

I disagree. I don't think this high-level convenience function should provide a giant security footgun in order to support one very rare use case. If that use case is important enough to get a high-level convenience function, then it's important enough to make that function actually convenient.

I'll leave it to @1st1 and @asvetlov decide whether they want to enforce this before merging.

1st1 replied:
> @asvetlov I've found the logic useful for some cases.

I'm also curious what were those cases.

> I'll leave it to @1st1 and @asvetlov decide whether they want to enforce this before merging.

After following this discussion I tend to side with Nathaniel's line of reasoning. The use-case for not setting `SO_REUSEADDR` is indeed extremely narrow. I strongly suggest to simply drop the `reuse_addr` parameter (and set `SO_REUSEADDR` always on *nix).

giampaolo replied:
> Again, my suggestion is: on Windows, use SO_EXCLUSIVEADDRUSE and don't use SO_REUSEADDR. That's all.

My current objections against using `SO_EXCLUSIVEADDRUSE` by default are:
1) [MS doc](https://docs.microsoft.com/en-us/windows/desktop/WinSock/using-so-reuseaddr-and-so-exclusiveaddruse) states <<Conversely, a socket with the SO_EXCLUSIVEADDRUSE set cannot necessarily be reused immediately after socket closure.>> which is precisely what we want to avoid here
2) setting no options at all already appears to match the same semantic of `SO_REUSEADDR` on UNIX (aka, return `EADDRINUSE` on second `bind()`). 
3) it's not being used by any major net framework I'm aware of and also by googling for `filetype:py SO_EXCLUSIVEADDRUSE` I get very few results. 
4) as for your script, I commented lines setting `SO_EXCLUSIVEADDRUSE` (23, 38, 47, 63) and on Windows 10 it prints exactly the same output. As such it appears it's probably not testing what it was intended for? Is it "stealing an already used port" the security scenario you're worried about? If so please paste a script able to reproduce it.

giampaolo replied:
I found a 6 years old issue where this was already discussed, basically leading to the same considerations/conclusions: 
https://bugs.python.org/issue8184

giampaolo replied:
> > I'll leave it to @1st1 and @asvetlov decide whether they want to enforce this before merging.

> After following this discussion I tend to side with Nathaniel's line of reasoning. The use-case for not setting `SO_REUSEADDR` is indeed extremely narrow. I strongly suggest to simply drop the `reuse_addr` parameter (and set `SO_REUSEADDR` always on *nix).

Yes, I now also agree this is probably not worth it. Also the fact that `reuse_addr` would have a different semantic on UNIX vs. Windows for UDP sockets is a bad sign.

njsmith replied:
>4. as for your script, I commented lines setting `SO_EXCLUSIVEADDRUSE` (23, 38, 47, 63) and on Windows 10 it prints exactly the same output. As such it appears it's probably not testing what it was intended for?

No, the point of the script is to demonstrate that settings `SO_EXCLUSIVEADDRUSE` doesn't break normal usage. It creates the situations where MSDN says using `SO_EXCLUSIVEADDRUSE` will cause problems, and demonstrates that in fact `SO_EXCLUSIVEADDRUSE` doesn't affect them at all. I haven't written a script to check that it actually fixes the bug/security hole it's supposed to fix. I guess that would be a good thing to do.

I believe the case where it's supposed to make a difference is:

Server A binds a socket. Then Server B attempts to bind the same address, and Server B uses `SO_REUSEADDR`. Supposedly, if Server A use `SO_EXCLUSIVEADDRUSE`, then Server B will get an error "socket already in use", but if Server A *doesn't* use `SO_EXCLUSIVEADDRUSE`, then Server B will hijack Server A's socket, and possibly cause other havoc. (The docs just say you get "undefined behavior".)

giampaolo replied:
> No, the point of the script is to demonstrate that settings `SO_EXCLUSIVEADDRUSE` doesn't break normal usage. It creates the situations where MSDN says using `SO_EXCLUSIVEADDRUSE` will cause problems, and demonstrates that in fact `SO_EXCLUSIVEADDRUSE` doesn't affect them at all.

If it doesn't affect them at all then why use it in the first place? It looks like the point of your script is to prove that MS doc is wrong. You are not even making a real case for it AFAICT.

> I believe the case where it's supposed to make a difference is: [this]

No. I've already explained why this is not the case multiple times: WinSock already do this by default. Why don't you verify your claims first?

```python
from socket import *
addr = ("localhost", 8888)
s1 = socket()
s1.bind(addr)
s2 = socket()
s2.setsockopt(SOL_SOCKET, SO_REUSEADDR, 1)  # will not cause port stealing
s2.bind(addr)  # fails with 10048 / address already in use
```

njsmith replied:
Wow. So now you're saying that the MSDN docs are wrong, and `SO_REUSEADDR` does not actually cause a security problem. That's super interesting if true! Thanks for pointing it out. However, what you are saying now is absolutely not what you said before. Quoting you:

> the semantic is different on Windows. When set to 1 it forces the socket to forcibly bind to a port in use by another socket, whereas on UNIX this results in EADDRINUSE error. It also sets the process in a state where it'll no longer respond to any signals or graceful kills, and must be forcibly killed (see test.support and BPO-2550).

> I think I changed my mind about disallowing to set `SO_REUSEADDR` on Windows because despite the security implications it appears there is a use-case for it. [→ so this is saying, you think that there are security implications]

This "already explained why this is not the case multiple times" thing you're claiming now is pure gaslighting.

> If it doesn't affect them at all then why use it in the first place? It looks like the point of your script is to prove that MS doc is wrong. You are not even making a real case for it AFAICT.

I have been straightforward and clear throughout about what I knew, what I suspected, and what my rationale was. I've responded to all of the points you've brought up, and repeated myself several times as you kept twisted what I said around into some unrelated strawman. And now you're being a patronizing jerk too.

I do care about the quality of the `socket` module, but I don't care enough to put up with your behavior. Have fun. Thanks for pointing out the `SO_REUSEADDR` thing; I'll probably investigate that more at some point. I'm unsubscribing from this thread now. Please don't tag me in again.

giampaolo replied:
@njsmith 
> Wow. So now you're saying that the MSDN docs are wrong, and `SO_REUSEADDR` does not actually cause a security problem. That's super interesting if true! Thanks for pointing it out.

That's not what I'm saying at all. We are all agreeing `SO_REUSEADDR` should not be used (nor allowed) on Windows because of the security implications. I even **quoted you** and agreed with you already that we shouldn't have a `reuse_addr` parameter anymore:
https://github.com/python/cpython/pull/11784#discussion_r256654410
So you either completely lost track of the conversation or you just want to provoke me. 

The point we're discussing about now is not `SO_REUSEADDR`. It's whether to use `SO_EXCLUSIVEADDRUSE` as *you* (not me) have been claiming we should do all along, because you erroneously think that if we don't, a second socket setting `SO_REUSEADDR` may steal the port of the first one (your words):

 > Server A binds a socket. Then Server B attempts to bind the same address, and Server B uses `SO_REUSEADDR`. Supposedly, if Server A use `SO_EXCLUSIVEADDRUSE`, then Server B will get an error "socket already in use", but if Server A _doesn't_ use `SO_EXCLUSIVEADDRUSE`, then Server B will hijack Server A's socket, and possibly cause other havoc. (The docs just say you get "undefined behavior".)

But the code I showed you [above](https://github.com/python/cpython/pull/11784#discussion_r256673038) (this):

```python
from socket import *
addr = ("localhost", 8888)
s1 = socket()
s1.bind(addr)
s2 = socket()
s2.setsockopt(SOL_SOCKET, SO_REUSEADDR, 1)  # will not cause port stealing
s2.bind(addr)  # correctly fails with 10048 / address already in use
```
...proves this is not the case and that you're wrong. This is what I've been trying to explain to you all along but never succeeded to. You've also never addressed my [arguments](https://github.com/python/cpython/pull/11784#discussion_r256575317) for **not** using `SO_EXCLUSIVEADDRUSE`, especially the number 1  where I pointed out that what MS doc says here is a problem:

<<Conversely, a socket with the SO_EXCLUSIVEADDRUSE set cannot necessarily be reused immediately after socket closure.>>

So again, not even an apparent lack of understanding on your part, but a refusal to fully engage in a rational debate and to answer my questions. Which can even be tolerable to some extent. When it stops being fine is the moment the discussion stops being technical and you start calling me names. Now this:

> This is a very odd conversation [...] Wow! That's super interesting if true! [...] This thing you're claiming now is pure gaslighting [...] And now you're being a **patronizing jerk** too. [...]  I'm unsubscribing from this thread now. [...] I don't care enough to put up with your behavior [...] Please don't tag me in again.

 ...is something I don't like and that I'm not willing to accept. Just so you know.

asvetlov commented on Lib/socket.py:
```diff
@@ -728,6 +728,78 @@ def create_connection(address, timeout=_GLOBAL_DEFAULT_TIMEOUT,
+        try:
+            sock = socket(af, socktype, proto)
+        except error as _:
+            err = _
```
I understand why do you need this code but please rename `_` to something more meaningful.
Usually, I assume that `_` is for unused var which is not the case here.

asvetlov commented on Lib/socket.py:
```diff
@@ -728,6 +728,78 @@ def create_connection(address, timeout=_GLOBAL_DEFAULT_TIMEOUT,
+            if reuse_addr:
+                try:
+                    sock.setsockopt(SOL_SOCKET, SO_REUSEADDR, 1)
+                except error:
```
When implicit error suppressing can be useful?

giampaolo replied:
Basically because if an exception has to occur I want it to happen on `bind()` later on, not here. 
In here the intent is just setting an option which refers to `bind()`.

asvetlov reviewed (commented):
I like the proposal but have a few comments regarding implementation.

giampaolo committed:
add 'reuse_addr' arg

giampaolo reviewed (commented):
added `reuse_addr` parameter

giampaolo committed:
rename method

giampaolo committed:
make family and type kw-only args

giampaolo committed:
raise ValueError if type is not SOCK_STREAM/DGRAM

giampaolo committed:
set IPV6_V6ONLY by default

giampaolo committed:
unittest: check IPV6_V6ONLY is set by default

vadmium commented on Doc/library/socket.rst:
```diff
@@ -595,6 +595,24 @@ The following functions all create :ref:`socket objects <socket-objects>`.
+    *backlog* is the queue size passed to :meth:`socket.listen` if
+    :data:`SOCK_STREAM` *type* is used.
+    *reuse_addr* and *reuse_port* dictates whether to use :data:`SO_REUSEADDR`
+    and :data:`SO_REUSEPORT` socket options.
```
What does _reuse_addr_=None mean? It looks like it enables REUSEADDR if supported, but doesn’t enforce it.

giampaolo replied:
Correct. One may use `reuse_addr_=False` to explicitly disable the option, but if left unspecified (`None`), what you want most of the times is to use SO_REUSEADDR. I took inspiration from `asyncio` which does the same.

giampaolo committed:
fix test failures

giampaolo committed:
adjust doc wording

giampaolo committed:
change var name

giampaolo committed:
set flags arg to None by default

giampaolo committed:
document that AF_INET is preferred if host's family is unclear

giampaolo committed:
introduce supports_hybrid_ipv46 and relative bind_socket arg

giampaolo committed:
various improvements:

- update doc examples
- force AF_INET6 if hybrid_ipv46=True
- add new functions to __all__
- add more tests

giampaolo committed:
update doc

giampaolo committed:
use 'localhost' in tests (safer)

giampaolo committed:
raise error on Windows if reuse_addr=True and type != SOCK_DGRAM

1st1 commented on Lib/socket.py:
```diff
@@ -728,6 +728,144 @@ def create_connection(address, timeout=_GLOBAL_DEFAULT_TIMEOUT,
+            or not hasattr(_socket, 'IPV6_V6ONLY'):
+        return False
+    try:
+        with socket(AF_INET6, SOCK_STREAM) as sock:
```
This check can be done once when this function is called for the first time and then cached, right?

giampaolo replied:
It can but IMO it's not worth it because:
1) this is a fast operation already: just instantiate a socket, set an option on it, close() it
2) in a server application this function is typically supposed to be called once (then the socket server will likely keep listening "forever")

Also, I'm wondering if there are platforms where `IPV6_V6ONLY` can be set globally, In that case we would probably want a fresh result.

gpshead replied:
Linux has `/proc/sys/net/ipv6/bindv6only` which can be changed at anytime...  http://man7.org/linux/man-pages/man7/ipv6.7.html

giampaolo replied:
Thanks. I will try experimenting with that to see it doesn't introduce issues or further differences.

giampaolo replied:
OK, this seems to determine the default value of `IPV6_V6ONLY` option for newly created sockets:

```
$ echo 0 > /proc/sys/net/ipv6/bindv6only
$ ./python -c "from socket import *; s = socket(AF_INET6); print(s.getsockopt(IPPROTO_IPV6, IPV6_V6ONLY))"
0
$ echo 1 > /proc/sys/net/ipv6/bindv6only
$ ./python -c "from socket import *; s = socket(AF_INET6); print(s.getsockopt(IPPROTO_IPV6, IPV6_V6ONLY))"
1
```

Even if `/proc/sys/net/ipv6/bindv6only` is set to 1, I am able to set it to 0 from Python, and unit-tests for `create_server` will pass. Basically this means that `/proc/sys/net/ipv6/bindv6only` does not disable the functionality. It just determines its default.

gpshead replied:
oh good. :)

1st1 commented on Doc/library/socket.rst:
```diff
@@ -595,6 +595,45 @@ The following functions all create :ref:`socket objects <socket-objects>`.
+   *reuse_addr* and *reuse_port* dictates whether to use :data:`SO_REUSEADDR`
+   and :data:`SO_REUSEPORT` socket options respectively.
+   *flags* is a bitmask for :meth:`getaddrinfo()`; if ``None``
+   :data:`AI_PASSIVE` is used.
```
Few suggested edits:

-----

A convenience function to create a server socket bound to *address*.  The function automates the following:

* resolving *address* with getaddrinfo
* finding the first resolved address that can be listened on
* automatically applying SO_REUSEADDR, if available
* disabling hybrid IPv4 / IPv6 listening
* calling "socket.listen()"

[add a "with-block" example here]

[list all parameters here]

giampaolo replied:
Mmm... IMO I think this would be a repetition of what I already describe when talking about the single parameters. I can probably add:

```
   Convenience function which creates a socket bound to *address* (a 2-tuple
   ``(host, port)``) and return the socket object upon which you can call
   :meth:`socket.accept()` in order to accept new connections.
+   It aims at automating all the necessary steps typically needed when creating
+   a server socket.

     [list all parameters here]

+     [add a "with-block" example here]
```

(also note I already added an example later in the doc)

1st1 replied:
The current text has grammatical errors, sentences are too long, and it is overall is a bit hard to follow, IMO.  But up to you, I guess.

giampaolo replied:
I changed the doc a bit by:
- adding a paragraph at the top: "Convenience function which aims at automating all the typical steps needed when creating a server socket"
- dividing each arg section with a new line

Does that make it a bit more clear? As for the grammatical errors, could you please point them out?

1st1 reviewed (commented):
Looks good overall, but please address my review.

giampaolo committed:
update doc + provide better error message on bind()

vstinner commented on Lib/socket.py:
```diff
@@ -728,6 +728,149 @@ def create_connection(address, timeout=_GLOBAL_DEFAULT_TIMEOUT,
+    except error:
+        return False
+
+def bind_socket(address, *, family=AF_UNSPEC, type=SOCK_STREAM, backlog=128,
```
I dislike hardcoded value backlog=128. I see that 'backlog' argument of sock.listen()  is now optional. Would it make sense to use backlog=None here and call sock.listen() with no parameter if it's None, or sock.listen(backlog) otherwise?

With the behavior that I proposed, I would be more comfortable with your changes replacing bind+listen (listen called with no parameter) with bind_socket.

Maybe raise an exception if backlog is set but socktype != SOCK_STREAM?

giampaolo replied:
The `backlog` default value topic is a tricky one. Some words have already been spent here:
https://github.com/python/cpython/pull/11215#discussion_r242762410

> Would it make sense to use backlog=None here and call sock.listen() with no parameter if it's None, or sock.listen(backlog) otherwise?

Yes, I think it would probably  be OK.

> Maybe raise an exception if backlog is set but socktype != SOCK_STREAM?

Mmmm... I think that would be a bit too strict/punitive.

vstinner commented on Lib/socket.py:
```diff
@@ -728,6 +728,149 @@ def create_connection(address, timeout=_GLOBAL_DEFAULT_TIMEOUT,
+    """
+    host, port = address
+    if host == "":
+        # https://mail.python.org/pipermail/python-ideas/2013-March/019937.html
```
Would you mind to write a short description, not just an url to the email?

vstinner commented on Lib/socket.py:
```diff
@@ -728,6 +728,149 @@ def create_connection(address, timeout=_GLOBAL_DEFAULT_TIMEOUT,
+    iswin = os.name in ('nt', 'cygwin')
+    if reuse_addr is None:
+        reuse_addr = not iswin and hasattr(_socket, 'SO_REUSEADDR')
+    elif reuse_addr and not hasattr(_socket, 'SO_REUSEADDR'):
```
I suggest to replace this "elif" and the following one with "if". It's more for the readability of the code: make sure that sanity tests are always run. (But the current code looks good.)

vstinner commented on Lib/socket.py:
```diff
@@ -728,6 +728,149 @@ def create_connection(address, timeout=_GLOBAL_DEFAULT_TIMEOUT,
+                try:
+                    sock.setsockopt(SOL_SOCKET, SO_REUSEADDR, 1)
+                except error:
+                    # Fail later on on bind(), for platforms which may not
```
typo: "on on".

vstinner commented on Doc/library/socket.rst:
```diff
@@ -595,6 +595,65 @@ The following functions all create :ref:`socket objects <socket-objects>`.
+   This parameter can be used in conjunction with :func:`supports_hybrid_ipv46`.
+
+   Here's an echo server example listening on all interfaces, port 8888,
+   and (possibly) hybrid IPv4/IPv6 support:
```
It uses TCP if I understand correctly. You may say it.

You use write "... support::" on a single line which is rendered as "... support:" and the code is formatted as you may expect ;-)

vstinner commented on Doc/library/socket.rst:
```diff
@@ -595,6 +595,65 @@ The following functions all create :ref:`socket objects <socket-objects>`.
+
+   Convenience function which aims at automating all the typical steps needed
+   when creating a server socket.
+   It creates a socket bound to *address* (a 2-tuple ``(host, port)``) and
```
It's only a 2-tuple for SOCK_STREAM, no? Maybe remove the comment in the parenthesis, since it's wrong for other socket families/types, no?

giampaolo replied:
No. This function only supports SOCK_STREAM and SOCK_DGRAM. Both of them need an `(host, port)` tuple as address.

vstinner replied:
@njsmith said something about SOCK_UNIX. Does it make any sense to support SOCK_UNIX as well?

giampaolo replied:
I think that would be better served as a separate function. Mostly because: 
1) the address would no longer be a `(host, ip)` tuple but a path
2) a UNIX socket may require it's own parameters (e.g. `mode`). It is true that we already have a parameter which applies to `SOCK_STREAM` sockets only (the `backlog`), but `SOCK_STREAM` are much more used than `AF_UNIX`.

vstinner commented on Doc/library/socket.rst:
```diff
@@ -595,6 +595,65 @@ The following functions all create :ref:`socket objects <socket-objects>`.
+   order to accept new connections.
+
+   Internally it relies on :meth:`getaddrinfo()` and returns the first socket
+   which can be bound to *address*.
```
Maybe add an empty line between each sentences to describe one parameter per paragraph? I'm not sure of the style in other documentations. Example: https://docs.python.org/dev/library/os.html#os.posix_spawn

vstinner commented on Doc/library/socket.rst:
```diff
@@ -595,6 +595,65 @@ The following functions all create :ref:`socket objects <socket-objects>`.
+   In this case the address returned by :meth:`socket.getpeername` when a new
+   IPv4 connection is accepted will be an IPv6 address represented as an
+   IPv4-mapped IPv6 address like ``":ffff:127.0.0.1"``.
+   If *hybrid_ipv46* is ``False`` it will explicitly disable this option on
```
nitpick: replace "is `False`" with "is false", since any object is accepted in practice. Same for "is ``True``" => "is true".

giampaolo replied:
Interesting. I didn't notice this was the convention in the doc. +1

vstinner commented on Lib/socket.py:
```diff
@@ -728,6 +728,149 @@ def create_connection(address, timeout=_GLOBAL_DEFAULT_TIMEOUT,
     else:
         raise error("getaddrinfo returns an empty list")
 
+def supports_hybrid_ipv46():
```
Would it make sense to add a type=SOCK_STREAM parameter?

giampaolo replied:
No because `hybrid_ipv46=True` doesn't work with SOCK_DGRAM (or any other proto type other than SOCK_STREAM AFAIK).

vstinner commented on Lib/socket.py:
```diff
@@ -728,6 +728,149 @@ def create_connection(address, timeout=_GLOBAL_DEFAULT_TIMEOUT,
+    info = getaddrinfo(host, port, family, type, 0, flags)
+    if family == AF_UNSPEC:
+        # prefer AF_INET over AF_INET6
+        info.sort(key=lambda x: x[0] == AF_INET, reverse=True)
```
maybe:

```suggestion
        info.sort(key=lambda x: x[0] != AF_INET)
```

vstinner commented on Lib/socket.py:
```diff
@@ -728,6 +728,149 @@ def create_connection(address, timeout=_GLOBAL_DEFAULT_TIMEOUT,
+            try:
+                sock.bind(sa)
+            except error as err:
+                es = '%s (while attempting to bind on address %r)' % \
```
Please rename es to err_msg.

vstinner commented on Lib/socket.py:
```diff
@@ -728,6 +728,149 @@ def create_connection(address, timeout=_GLOBAL_DEFAULT_TIMEOUT,
+
+    err = None
+    for res in info:
+        af, socktype, proto, canonname, sa = res
```
may rename "sa" to "sock_addr" or "address".

giampaolo replied:
This is the same name used in `create_connection` and also in socket module's doc example. Better we stay consistent.

vstinner commented on Lib/socket.py:
```diff
@@ -728,6 +728,149 @@ def create_connection(address, timeout=_GLOBAL_DEFAULT_TIMEOUT,
+                    # consistent across different platforms.
+                    try:
+                        sock.setsockopt(IPPROTO_IPV6, IPV6_V6ONLY, 1)
+                    except NameError:
```
Please add a comment explaining that IPPROTO_IPV6 or IPV6_V6ONLY is not available. I was surprised by "except NameError:" here at the first read.

vstinner commented on Lib/socket.py:
```diff
@@ -728,6 +728,149 @@ def create_connection(address, timeout=_GLOBAL_DEFAULT_TIMEOUT,
+                raise error(err.errno, es) from None
+            if socktype == SOCK_STREAM:
+                sock.listen(backlog)
+            # Break explicitly a reference cycle.
```
*maybe* move these 3 lines in an "else:" block.

vstinner commented on Lib/socket.py:
```diff
@@ -728,6 +728,149 @@ def create_connection(address, timeout=_GLOBAL_DEFAULT_TIMEOUT,
+
+    if err is not None:
+        raise err
+    else:
```
maybe remove the "else:" to put the raise at the same level than if, to show that it is the final instruction if nothing else has been executed previously.

vstinner commented on Lib/test/test_asyncio/functional.py:
```diff
@@ -60,21 +60,13 @@ def tcp_server(self, server_prog, *,
 
-        sock = socket.socket(family, socket.SOCK_STREAM)
-
+        sock = socket.bind_socket(addr, family=family, backlog=backlog)
```
Previously, the timeout was set before bind+listen. Now it's set after. Do you know if bind or listen are impacted by the timeout? (I don't think so.)

giampaolo replied:
No. Same effect. `accept()` is the one affected by the timeout but setting it before or after is the same.

vstinner commented on Doc/whatsnew/3.8.rst:
```diff
@@ -214,6 +214,15 @@ contain characters unrepresentable at the OS level.
+------
+
+Added :meth:`~socket.bind_socket()` and :meth:`~socket.supports_hybrid_ipv46()`
+convenience functions to automatize the necessary tasks usually involved when
```
You may name tasks:

```suggestion
convenience functions to automatize the necessary tasks (bind, listen) usually involved when
```

gpshead replied:
"automatize" isn't a word. :)  I'd suggest "to automate" or "that encapsulate" perhaps.

giampaolo replied:
False friend betrayed me. =)
Fixed.

vstinner commented on Doc/library/socket.rst:
```diff
@@ -595,6 +595,65 @@ The following functions all create :ref:`socket objects <socket-objects>`.
    .. versionchanged:: 3.2
       *source_address* was added.
 
+.. function:: bind_socket(address, *, family=AF_UNSPEC, type=SOCK_STREAM, backlog=128, reuse_addr=None, reuse_port=False, flags=None, hybrid_ipv46=False)
```
bikeshedding on the function name... since listen can also be involved, maybe reuse asyncio name: "create_server()"?

In asyncio, it's more complex than "just" create a socket + bind + listen, the "server" really handles incoming connections, etc.

But I still like "server" name since I'm always confused by the socket API, I never recall how to use it, and "server" is an obvious hint and it can be used to listen for incoming *client* connections.

Well, it's just a suggestion :-)

vstinner reviewed (commented):
I like the bind_socket function! A first round of review. I don't see any major issues, it's mostly a long list of suggestions.

vstinner commented on Lib/test/test_asyncio/test_streams.py:
```diff
@@ -796,10 +794,7 @@ def test_drain_raises(self):
-            with sock:
-                sock.bind(('localhost', 0))
-                sock.listen(1)
+            with socket.bind_socket(('localhost', 0)) as sock:
```
I would prefer to keep an explicit backlog=1... I'm not sure why listen(1) was used, so in case of doubt, I prefer to not change it :-)

giampaolo replied:
I'm pretty sure it makes no difference. Basically all other tests do not use the backlog.

giampaolo committed:
rename bind_socket() to create_server()

giampaolo committed:
get rid of reuse_addr arg

giampaolo committed:
address @vstinner comments

jaraco commented on Lib/socket.py:
```diff
@@ -728,6 +728,140 @@ def create_connection(address, timeout=_GLOBAL_DEFAULT_TIMEOUT,
+        family = AF_INET6
+    info = getaddrinfo(host, port, family, type, 0, flags)
+    if family == AF_UNSPEC:
+        # prefer AF_INET over AF_INET6
```
I'm strongly opposed to this choice. If the IP stack is giving preference to IPv6 by presenting it first, Python should not be overriding it to prefer the less capable legacy approach. I see this approach as being comparable to preferring Python 2 syntax in a compatibility library. The systems should be designed to prefer the newer, more advanced behavior over the older one, even if the older is dominant. Giving preference to IPv4 leaves technical debt that eventually will need to be rectified (with all of the transitional problems that brings).

In other words, I believe this choice precludes the ability to solve my primary use-case for this function, which is:

"Create a socket listening on the best available protocol" where best available is IPv6+IPv4 or IPv6 or IPv4 in descending order.

giampaolo replied:
The use case you're describing is already covered by:
```python
create_server(addr, hybrid_ipv46=support_hybrid_ipv46())
```

jaraco replied:
Okay. I'd like to run some tests.

jaraco replied:
Okay. I confirm this usage behaves the way I would expect. I suppose I could accept this approach, if necessary.

```
>>> socket.create_server((None, 0), hybrid_ipv46=socket.supports_hybrid_ipv46())
<socket.socket fd=6, family=AddressFamily.AF_INET6, type=SocketKind.SOCK_STREAM, proto=6, laddr=('::', 53712, 0, 0)>
>>> socket.create_server(('localhost', 0), hybrid_ipv46=socket.supports_hybrid_ipv46())
<socket.socket fd=7, family=AddressFamily.AF_INET6, type=SocketKind.SOCK_STREAM, proto=6, laddr=('::1', 53713, 0, 0)>
```

I'd have preferred this behavior would be the default, as that's the default server behavior I've used on other IP servers I've written. I find it awkward that the default is "prefer the legacy behavior", especially for a new function for which there's no prior expectation. At what point do you think Python would want to change this default behavior to prefer IPv6?

giampaolo committed:
fix NameError + add comment clarifying why we in case of ambiguous host we prever AF_INET over AF_INET6

giampaolo committed:
fix test failures

vstinner commented on Doc/library/socket.rst:
```diff
@@ -622,12 +622,12 @@ The following functions all create :ref:`socket objects <socket-objects>`.
    :data:`AI_PASSIVE` is used.
 
-   If *hybrid_ipv46* is ``True`` and the platform supports it the socket will
+   If *hybrid_ipv46* is tre and the platform supports it the socket will
```
typo: true

giampaolo committed:
don't use getaddrinfo(), change function signature

https://github.com/python/cpython/pull/11784#issuecomment-463736515

giampaolo committed:
rename arg hybrid_ipv46 -> dualstack_ipv6

giampaolo committed:
update doc

giampaolo committed:
fix ftplib bug, skip IPV6 tests

terryjreedy committed:
idlelib/rpc.py: add missing address argument

Omission caused IDLE to crash when started.

terryjreedy reviewed (commented):
PR branches should generally be branches in one personal fork of cpython, not in the main repository.

This PR completely broke IDLE.  Anyone touching IDLE code must test manually in addition to running the half-complete test suite.  After adding the missing address parameter, I did enough testing to think that the result is likely OK.  But I would like someone who is familiar with sockets to review that the new create_server call is equivalent to the code it replaces.

giampaolo committed:
update doc

giampaolo committed:
fix doc example + remove UDP-related test code which is no longer used

giampaolo committed:
merge from master

gpshead commented on Doc/library/socket.rst:
```diff
@@ -1779,6 +1824,43 @@ sends traffic to the first one connected successfully. ::
+
+   HOST = ""
+   PORT = 50007
+   if socket.has_dualstack_ipv6():
```
What does `socket.has_dualstack_ipv6()` return on an IPv6-only host?  If it returns False in that situation, this code would fail on IPv6 only hosts as it'll attempt the default family=AF_INET path.

Your example code repeats these same four lines everywhere and suggests that people use them.  I think we should encapsulate this idiom _into_ the create_server API itself including working in IPv6 only situations.

giampaolo replied:
> What does `socket.has_dualstack_ipv6()` return on an IPv6-only host? 

False

> If it returns False in that situation, this code would fail on IPv6 only hosts as it'll attempt the default family=AF_INET path.

Correct. As it stands, in order to bind on IPv6 + dual stack you have to specify both arguments:

```python
>>> create_server(addr, family=AF_INET6, dualstack_ipv6=True)
```

> Your example code repeats these same four lines everywhere and suggests that people use them.

I added an extra example mainly because the [examples section](https://docs.python.org/3/library/socket.html#example) already iterated over echo client/server samples and didn't mention `create_connection`, but I agree this is probably too much information. I removed it.

gpshead commented on Doc/library/socket.rst:
```diff
@@ -595,6 +595,51 @@ The following functions all create :ref:`socket objects <socket-objects>`.
+   object.
+
+   *family* should be either :data:`AF_INET` or :data:`AF_INET6`.
+   *backlog* is the queue size passed to :meth:`socket.listen`.
```
Describe what backlog=None means.

giampaolo replied:
Good idea. Updated doc (including your suggestion of making it default to 0 instead of None):
```
   *backlog* is the queue size passed to :meth:`socket.listen`; when ``0``
   a default reasonable value is chosen.
```

gpshead commented on Doc/library/socket.rst:
```diff
@@ -595,6 +595,51 @@ The following functions all create :ref:`socket objects <socket-objects>`.
+   *backlog* is the queue size passed to :meth:`socket.listen`.
+   *reuse_port* dictates whether to set :data:`SO_REUSEPORT` socket option.
+
+   If *dualstack_ipv6* is true and the platform supports it the socket will
```
Describe what happens if this is true but the platform doesn't support it.

giampaolo replied:
Done (added "will raise ValueError").

gpshead commented on Doc/library/socket.rst:
```diff
@@ -595,6 +595,51 @@ The following functions all create :ref:`socket objects <socket-objects>`.
+   If *dualstack_ipv6* is false it will explicitly disable this functionality
+   on platforms that enable it by default (e.g. Linux).
+   For platforms not supporting this functionality natively you could use this
+   `MultipleSocketsListener recipe <http://code.activestate.com/recipes/578504/>`__.
```
Use an https link.  (also, activestate appears to be down or DOSed right now)

giampaolo replied:
Good catch. Done.

gpshead commented on Doc/library/socket.rst:
```diff
@@ -595,6 +595,51 @@ The following functions all create :ref:`socket objects <socket-objects>`.
+.. function:: has_dualstack_ipv6()
+
+   Return ``True`` if the platform supports creating a :data:`SOCK_STREAM`
+   socket which can handle both :data:`AF_INET` or :data:`AF_INET6`
```
Describe how one creates such a socket when this returns `True` given that the `socket()` call only accepts a single `family`.

giampaolo replied:
This is already described in `create_function` which is right above (and ends with a code sample). I don't think it's worth repeating the description of the behavior (or showing the same code sample).

gpshead replied:
I was assuming someone will want to use this in combination with plain `socket()` calls without `create_socket()` at times.  ie: Should we mention the setsockopt IPV6_V6ONLY 0 dance for people not using create_socket?  Or just leave that as one for the networking books?

giampaolo replied:
I'd leave it out as a low-level detail. As of now, the only thing that the doc mentions is the existence of `IPV6_*` constants. We may decide to improve that as a separate doc fix though.

vadmium replied:
I think AF_INET and AF_INET6 are mutually exclusive for a given socket. Perhaps it is more accurate to say the socket can handle both IP v4 and v6.

gpshead commented on Doc/library/socket.rst:
```diff
@@ -595,6 +595,51 @@ The following functions all create :ref:`socket objects <socket-objects>`.
    .. versionchanged:: 3.2
       *source_address* was added.
 
+.. function:: create_server(address, *, family=AF_INET, backlog=None, reuse_port=False, dualstack_ipv6=False)
```
Creating a new API in 2019 that defaults to IPv4 without dualstack seems backwards. People will use these defaults without realizing that they're shutting out a big part of the world.

Our defaults should require users to opt-out of dualstack IPv6 and that the dualstack parameter should perhaps be a tri-state rather than a bool: 'no', 'required', 'allow'.  the default I'd look for would be allow.  such that it would not be an error on a mythical non-dualstack capable host.

When dualstack is enabled, does the family= value have a useful meaning?

On non-dualstack-socket capable hosts (what are those even?) where the user needs a server listening on _both_ IPv4 and IPv6, isn't it true that they need to create _two server sockets_?  `create_server()` doesn't really have a way to address that case.  Supporting such a situation is going to require a `has_dualstack_ipv6()` conditional and wrangling multiple sockets no matter what.

For a situation where people just want a single "most capable" server socket, your four line conditional code seems like it belongs inside of `create_server` itself.

giampaolo replied:
> Creating a new API in 2019 that defaults to IPv4 without dualstack seems backwards. People will use these defaults without realizing that they're shutting out a big part of the world.

The rationale for having dual-stack disabled by default IMO are 2, consistency and desirableness. In details:
- not every platform supports dual-stack, so enabling it by default seems inconsistent 
- not all platforms that support it have it enabled by default. E.g. on Linux and FreeBSD it's enabled, on OSX it's not.
-  in case of IPv4 connection, `getpeername()` will return an IPv4 mapped address like `'::ffff:127.0.0.1'`, which is surprising (more [here](https://github.com/python/cpython/pull/11784/#issuecomment-463678671)). To me this is not desirable regardless of how much spread IPv6 adoption is, or even if dual-stack was supported on all platforms. Not all users who want IPv6 necessarily want the dual stack (honestly, initially I was even tempted to exclude this option completely because of all this platform differences - and I'm still happy to take that into consideration BTW). 

Also note that asyncio disables dual-stack by default, basically for the same reasons. asyncio case is different though. It does not expose an option to enable dual-stack because it would be useless: it is already able to serve both families with 2 distinct sockets which are handled by the IO loop. We cannot do that here though, unless we use the external recipe I mentioned in the doc (and which should not end up in socket.py IMO).

> Our defaults should require users to opt-out of dualstack IPv6 and that the dualstack parameter should perhaps be a tri-state rather than a bool: 'no', 'required', 'allow'. the default I'd look for would be allow. such that it would not be an error on a mythical non-dualstack capable host.

I hear you. I also considered that, but I couldn't come up with an API which made sense. Basically that's why the API evolved into being "explicit" by forcing you to specify both *family* and *dualstack_ipv6*.  

> On non-dualstack-socket capable hosts (what are those even?)

Windows does not support dual-stack. Linux, OSX and FreeBSD do. I updated the doc adding:
"Most POSIX platforms are supposed to support this option."

> When dualstack is enabled, does the family= value have a useful meaning?

As this currently stands, *family* has to be explicitly set to `AF_INET6`. One may argue that should be implicit if `dualstack_ipv6=True`. As of now I'm more inclined to think it should be set explicitly because:
- it helps clarifying that you're primarily dealing with an IPv6 socket, even if it has mixed capabilities. 
- it reflects the fact that `socket.socket()` default family is `AF_INET`, so the same way you do `socket.socket(AF_INET6)` you should do `socket.create_server(addr, family=AF_INET6)`. 

But I agree this is indeed debatable and I have mixed feelings about it.

> On non-dualstack-socket capable hosts (what are those even?) where the user needs a server listening on _both_ IPv4 and IPv6, isn't it true that they need to create _two server sockets_? `create_server()` doesn't really have a way to address that case. 

Correct. If dual-stack is not supported natively by the kernel you have to create 2 sockets. The ActiveState recipe I mentioned in the doc includes a class which does that. The reason for not including it in socket.py is because the API is non-obvious / incompatible. E.g. `fileno()`, `detach()`, `family` and possibly others wouldn't know what to return.

gpshead commented on Lib/socket.py:
```diff
@@ -728,6 +728,89 @@ def create_connection(address, timeout=_GLOBAL_DEFAULT_TIMEOUT,
+        if not has_dualstack_ipv6():
+            raise ValueError("dualstack_ipv6 not supported on this platform")
+        if family != AF_INET6:
+            raise ValueError("dualstack_ipv6 requires AF_INET6 family")
```
It is ugly as an API to force someone to pass two different flags when they want to enable a single thing.  this reinforces my thought that an explicit family and dualstack_ipv6 setting are too low level for a create_server() API.

giampaolo replied:
Agree with you this is debatable although I currently have no solution.

gpshead replied:
you've convinced me.  `socket` is still a low level module, more abstraction gets into the land of higher level networking frameworks of which there are plenty built on top of all of this. :)

gpshead commented on Lib/socket.py:
```diff
@@ -728,6 +728,89 @@ def create_connection(address, timeout=_GLOBAL_DEFAULT_TIMEOUT,
+            msg = '%s (while attempting to bind on address %r)' % \
+                (err.strerror, address)
+            raise error(err.errno, msg) from None
+        sock.listen(backlog or 0)
```
why not just set the backlog default value to 0?

giampaolo replied:
Yes, I think that makes sense. Done.

giampaolo committed:
set backlog=0 instead of None; address doc-related review comments

giampaolo reviewed (commented):
@gpshead thanks for your feedback. I added some comments inline.

giampaolo committed:
change unit-tests so that client does not rely on getaddrinfo() - better be explicit what family the client is supposed to use

giampaolo committed:
fix wrong client addr + fix travis test due to extra whitespace

gpshead committed:
automatize -> automate

vadmium commented on Doc/library/socket.rst:
```diff
@@ -595,6 +595,53 @@ The following functions all create :ref:`socket objects <socket-objects>`.
 
+.. function:: create_server(address, *, family=AF_INET, backlog=0, reuse_port=False, dualstack_ipv6=False)
+
+   Convenience function which creates a :data:`SOCK_STREAM` type socket
```
Any reason why you don't just say it creates a TCP socket? I think this is guaranteed by Posix for AF_INET and AF_INET6 with SOCK_STREAM and _proto_ set to zero.

vadmium commented on Doc/library/socket.rst:
```diff
@@ -595,6 +595,53 @@ The following functions all create :ref:`socket objects <socket-objects>`.
+         s = socket.create_server(addr)
+
+   .. note::
+    On POSIX platforms :data:`SO_REUSEADDR` socket option is set in order to
```
Add _the_: On Posix platforms, the SO_REUSEADDR socket option . . .

vadmium commented on Lib/idlelib/rpc.py:
```diff
@@ -530,9 +530,8 @@ class RPCClient(SocketIO):
-        self.listening_sock.bind(address)
-        self.listening_sock.listen(1)
+        self.listening_sock = socket.create_server(
+            address, family=family, type=type, backlog=1)
```
Is _type_ a valid keyword? If not, maybe this needs a test case :)

giampaolo replied:
Ouch! I missed that, thanks. I just removed this part and left idlelib alone in f786884. I will probably get back at this in a separate PR because it's more controversial.

vadmium commented on Lib/socket.py:
```diff
@@ -728,6 +728,89 @@ def create_connection(address, timeout=_GLOBAL_DEFAULT_TIMEOUT,
+            raise error(err.errno, msg) from None
+        sock.listen(backlog)
+        return sock
+    except Exception:
```
Use plain _except_ to catch all exceptions (e.g. KeyboardInterrupt could be raised during a slow DNS lookup)

giampaolo replied:
Good point. I think it's saner to just catch `socket.error`, the same way it's done in `create_connection()`. Changed it in f786884.

vadmium commented on Lib/test/test_socket.py:
```diff
@@ -6053,9 +6053,133 @@ def test_new_tcp_flags(self):
+            with sock:
+                conn, _ = sock.accept()
+                with conn:
+                    event.wait(self.timeout)
```
Is this event necessary? It is set just after the thread is started.

giampaolo replied:
Yes it is because you want to wait for `accept()` to succeed before doing `connect()`. Without this sync primitive in place the test will fail.

giampaolo committed:
Merge branch 'master' into bind-socket

giampaolo committed:
update doc; catch socket.error instead of Exception; remove idlelib integration

giampaolo committed:
Merge branch 'master' into bind-socket

giampaolo committed:
update doc (remove ref to activestate recipe)
